File: src/synapt_eval/report_card/markdown.py
"""Markdown report card renderer."""

from __future__ import annotations

from synapt_eval.report_card.types import (
    CategorySection,
    ReportCard,
    ReportCardFooter,
    ReportCardHeader,
    TrendingEntry,
)
from synapt_eval.suggestion_engine.types import Suggestion


def generate_markdown(report_card: ReportCard) -> str:
    """Render a ReportCard as a markdown string."""
    parts: list[str] = []

    parts.append(_render_header(report_card.header))

    for section in report_card.sections:
        parts.append(_render_section(section))

    if report_card.trending:
        parts.append(_render_trending(report_card.trending))

    if report_card.suggestions:
        parts.append(_render_suggestions(report_card.suggestions))

    parts.append(_render_footer(report_card.footer))

    return "\n".join(parts)


def _render_header(header: ReportCardHeader) -> str:
    lines = ["# Eval Report Card", ""]

    lines.append(f"**Run ID**: {header.run_id}  ")
    lines.append(f"**Timestamp**: {header.timestamp}  ")
    if header.commit:
        lines.append(f"**Commit**: {header.commit}  ")
    lines.append(
        f"**Fixtures**: {header.fixture_count} across {header.category_count} categories  "
    )

    severity_parts = []
    for level in ("error", "warning", "info"):
        count = header.severity_counts.get(level, 0)
        if count > 0:
            severity_parts.append(f"{count} {level.upper()}")
    if severity_parts:
        lines.append(f"**Severity**: {' / '.join(severity_parts)}")

    lines.append("")
    lines.append("---")
    lines.append("")
    return "\n".join(lines)


def _render_section(section: CategorySection) -> str:
    lines = [f"## {section.category}", ""]

    lines.append("| Metric | Value |")
    lines.append("|--------|-------|")
    lines.append(f"| P@5 | {section.metrics.p_at_5:.3f} |")
    lines.append(f"| R@10 | {section.metrics.r_at_10:.3f} |")
    if section.metrics.tau is not None:
        lines.append(f"| Tau | {section.metrics.tau:.3f} |")
    lines.append(f"| N | {section.metrics.n} |")
    lines.append("")

    if section.suggestions:
        lines.append("### Suggestions")
        lines.append("")
        for s in section.suggestions:
            level = s.severity.level.upper()
            lines.append(f"- [{level}] **{s.rule_name}**: {s.message}")
            if s.fix_hint:
                lines.append(f"  - *Fix*: {s.fix_hint}")
        lines.append("")

    lines.append("---")
    lines.append("")
    return "\n".join(lines)


def _render_trending(entries: list[TrendingEntry]) -> str:
    if not entries:
        return ""

    categories = sorted({e.category for e in entries})
    lines = ["## Trending", ""]

    for category in categories:
        cat_entries = [e for e in entries if e.category == category]
        if not cat_entries:
            continue

        lines.append(f"### {category}")
        lines.append("")

        metric_names = sorted({k for e in cat_entries for k in e.metrics})
        header_cols = ["Run"] + [m.upper() for m in metric_names]
        lines.append("| " + " | ".join(header_cols) + " |")
        lines.append("| " + " | ".join("---" for _ in header_cols) + " |")

        for entry in cat_entries:
            row = [entry.run_id]
            for m in metric_names:
                val = entry.metrics.get(m)
                row.append(f"{val:.3f}" if val is not None else "-")
            lines.append("| " + " | ".join(row) + " |")

        lines.append("")

    lines.append("---")
    lines.append("")
    return "\n".join(lines)


def _render_suggestions(suggestions: list[Suggestion]) -> str:
    lines = ["## Suggestions Summary", ""]
    lines.append("| # | Severity | Rule | Category | Message |")
    lines.append("|---|----------|------|----------|---------|")

    for i, s in enumerate(suggestions, 1):
        level = s.severity.level.upper()
        category = s.category or "-"
        lines.append(f"| {i} | {level} | {s.rule_name} | {category} | {s.message} |")

    lines.append("")
    lines.append("---")
    lines.append("")
    return "\n".join(lines)


def _render_footer(footer: ReportCardFooter) -> str:
    lines = ["## Result", ""]

    status = "PASSED" if footer.passed else "FAILED"
    parts = []
    if footer.regression_summary:
        parts.append(footer.regression_summary)
    if footer.total_suggestions > 0:
        parts.append(f"{footer.total_suggestions} suggestion(s) for improvement")
    else:
        parts.append("No suggestions")

    detail = ". ".join(parts) + "." if parts else ""
    lines.append(f"**{status}** -- {detail}")

    if footer.deltas:
        lines.append("")
        lines.append("### Regression Deltas")
        lines.append("")
        lines.append("| Category | Metric | Baseline | Current | Delta |")
        lines.append("|----------|--------|----------|---------|-------|")
        for d in footer.deltas:
            flag = " [REGRESSION]" if d.regression else ""
            lines.append(
                f"| {d.category} | {d.metric} | "
                f"{d.baseline:.3f} | {d.current:.3f} | "
                f"{d.delta:+.3f}{flag} |"
            )
        lines.append("")

    lines.append("")
    return "\n".join(lines)

File: ts/src/report_card/markdown.ts
import type { Suggestion } from "../suggestion_engine/types.ts";
import type {
  CategorySection,
  ReportCard,
  ReportCardFooter,
  ReportCardHeader,
  TrendingEntry,
} from "./types.ts";

export function generateMarkdown(reportCard: ReportCard): string {
  const parts: string[] = [];

  parts.push(renderHeader(reportCard.header));

  for (const section of reportCard.sections) {
    parts.push(renderSection(section));
  }

  if (reportCard.trending.length > 0) {
    parts.push(renderTrending(reportCard.trending));
  }

  if (reportCard.suggestions.length > 0) {
    parts.push(renderSuggestions(reportCard.suggestions));
  }

  parts.push(renderFooter(reportCard.footer));

  return parts.join("\n");
}

function renderHeader(header: ReportCardHeader): string {
  const lines = ["# Eval Report Card", ""];
  lines.push(`**Run ID**: ${header.runId}  `);
  lines.push(`**Timestamp**: ${header.timestamp}  `);
  if (header.commit) {
    lines.push(`**Commit**: ${header.commit}  `);
  }
  lines.push(
    `**Fixtures**: ${header.fixtureCount} across ${header.categoryCount} categories  `,
  );

  const severityParts: string[] = [];
  for (const level of ["error", "warning", "info"]) {
    const count = header.severityCounts[level] ?? 0;
    if (count > 0) {
      severityParts.push(`${count} ${level.toUpperCase()}`);
    }
  }
  if (severityParts.length > 0) {
    lines.push(`**Severity**: ${severityParts.join(" / ")}`);
  }

  lines.push("", "---", "");
  return lines.join("\n");
}

function renderSection(section: CategorySection): string {
  const lines = [`## ${section.category}`, ""];
  lines.push("| Metric | Value |");
  lines.push("|--------|-------|");
  lines.push(`| P@5 | ${section.metrics.pAt5.toFixed(3)} |`);
  lines.push(`| R@10 | ${section.metrics.rAt10.toFixed(3)} |`);
  if (section.metrics.tau != null) {
    lines.push(`| Tau | ${section.metrics.tau.toFixed(3)} |`);
  }
  lines.push(`| N | ${section.metrics.n} |`);
  lines.push("");

  if (section.suggestions.length > 0) {
    lines.push("### Suggestions", "");
    for (const s of section.suggestions) {
      const level = s.severity.level.toUpperCase();
      lines.push(`- [${level}] **${s.ruleName}**: ${s.message}`);
      if (s.fixHint) {
        lines.push(`  - *Fix*: ${s.fixHint}`);
      }
    }
    lines.push("");
  }

  lines.push("---", "");
  return lines.join("\n");
}

function renderTrending(entries: TrendingEntry[]): string {
  if (entries.length === 0) return "";

  const categories = [...new Set(entries.map((e) => e.category))].sort();
  const lines = ["## Trending", ""];

  for (const category of categories) {
    const catEntries = entries.filter((e) => e.category === category);
    if (catEntries.length === 0) continue;

    lines.push(`### ${category}`, "");

    const metricNames = [
      ...new Set(catEntries.flatMap((e) => Object.keys(e.metrics))),
    ].sort();
    const headerCols = ["Run", ...metricNames.map((m) => m.toUpperCase())];
    lines.push("| " + headerCols.join(" | ") + " |");
    lines.push("| " + headerCols.map(() => "---").join(" | ") + " |");

    for (const entry of catEntries) {
      const row = [entry.runId];
      for (const m of metricNames) {
        const val = entry.metrics[m];
        row.push(val != null ? val.toFixed(3) : "-");
      }
      lines.push("| " + row.join(" | ") + " |");
    }
    lines.push("");
  }

  lines.push("---", "");
  return lines.join("\n");
}

function renderSuggestions(suggestions: Suggestion[]): string {
  const lines = ["## Suggestions Summary", ""];
  lines.push("| # | Severity | Rule | Category | Message |");
  lines.push("|---|----------|------|----------|---------|");

  for (let i = 0; i < suggestions.length; i++) {
    const s = suggestions[i];
    const level = s.severity.level.toUpperCase();
    const category = s.category ?? "-";
    lines.push(
      `| ${i + 1} | ${level} | ${s.ruleName} | ${category} | ${s.message} |`,
    );
  }

  lines.push("", "---", "");
  return lines.join("\n");
}

function renderFooter(footer: ReportCardFooter): string {
  const lines = ["## Result", ""];

  const status = footer.passed ? "PASSED" : "FAILED";
  const parts: string[] = [];
  if (footer.regressionSummary) {
    parts.push(footer.regressionSummary);
  }
  if (footer.totalSuggestions > 0) {
    parts.push(`${footer.totalSuggestions} suggestion(s) for improvement`);
  } else {
    parts.push("No suggestions");
  }

  const detail = parts.length > 0 ? parts.join(". ") + "." : "";
  lines.push(`**${status}** -- ${detail}`);

  if (footer.deltas.length > 0) {
    lines.push("", "### Regression Deltas", "");
    lines.push("| Category | Metric | Baseline | Current | Delta |");
    lines.push("|----------|--------|----------|---------|-------|");
    for (const d of footer.deltas) {
      const flag = d.regression ? " [REGRESSION]" : "";
      const sign = d.delta >= 0 ? "+" : "";
      lines.push(
        `| ${d.category} | ${d.metric} | ${d.baseline.toFixed(3)} | ${d.current.toFixed(3)} | ${sign}${d.delta.toFixed(3)}${flag} |`,
      );
    }
    lines.push("");
  }

  lines.push("");
  return lines.join("\n");
}
